### PR TITLE
Check for FENV_ACCESS support, conditional checks in math_errorhandling

### PR DESCRIPTION
**Pull request by domin144**
Added 'fenv_access_supported' compiler check to detect compilers, which do not support "#pragma STDC FENV_ACCESS ON".

Removed misleading LONG_DOUBLE_EXCEPTION_TEST and DOUBLE_EXCEPTION_TEST defines. These both took the current value of EXCEPTION_TEST not the one from the assignment time. Both `long double` and `double` tests were run, even when PICOLIBC_LONG_DOUBLE_NOEXCEPT or PICOLIBC_DOUBLE_NOEXCEPT were defined. As the tests passed anyway and the exceptions were intentionally not set, the tests are left enabled.

### CODE REVIEW

**Comment by arichardson** [comment]
For `long double` and other softfloat implementations a lot of this depends on the compiler builtins library (libgcc). LLVM's compiler-rt does not currently raise most exceptions for softfloat routines. I have a patch that fixes most of this (but is still missing overflow/underflow flags: https://reviews.llvm.org/D98332

I think the only way to check for this would be a runtime behaviour check in the test?

**Comment by keith-packard** [comment]
> For `long double` and other softfloat implementations a lot of this depends on the compiler builtins library (libgcc). LLVM's compiler-rt does not currently raise most exceptions for softfloat routines. I have a patch that fixes most of this (but is still missing overflow/underflow flags: https://reviews.llvm.org/D98332

libgcc doesn't support exceptions for softfloat either. Where do you store exception information on processors which have no FPU? Or are you only interested in supporting exceptions for processors that have a limited FPU with support for only smaller datatypes?

> I think the only way to check for this would be a runtime behaviour check in the test?

That's a really good idea. Do you have an idea on how we can detect support at runtime? Maybe just open-code a basic exception (0/0 or similar) and see if that sets an exception flag?

**Comment by arichardson** [comment]
> > For `long double` and other softfloat implementations a lot of this depends on the compiler builtins library (libgcc). LLVM's compiler-rt does not currently raise most exceptions for softfloat routines. I have a patch that fixes most of this (but is still missing overflow/underflow flags: https://reviews.llvm.org/D98332
> 
> libgcc doesn't support exceptions for softfloat either. Where do you store exception information on processors which have no FPU? Or are you only interested in supporting exceptions for processors that have a limited FPU with support for only smaller datatypes?
> 
My use case there was getting soft-float 128-bit exceptions to match the FPU ones for float/double (as that's what the freebsd testsuite expects - it was designed for x86 long double).

> > I think the only way to check for this would be a runtime behaviour check in the test?
> 
> That's a really good idea. Do you have an idea on how we can detect support at runtime? Maybe just open-code a basic exception (0/0 or similar) and see if that sets an exception flag?

Exactly that's what I was thinking. I think 0/0 would be a good example (or subtraction of infinities).

**Comment by keith-packard** [comment]
> Exactly that's what I was thinking. I think 0/0 would be a good example (or subtraction of infinities).

You'll want to use global volatile variables holding the values or clang will probably optimize this away. In any case, this seems like a good direction, avoiding the complexity of trying to figure this out in the build system.

**Comment by domin144** [comment]
I was testing clang + compiler_rt with this configuration: `--target=arm-none-eabihf -mfloat-abi=hard -march=armv7em -mfpu=fpv4-sp-d16`. I was debugging math_errhandling test. Clang **was producing** an exception even though it was **not expected to**. My initial diagnosis was, that in my particular case clang was calculating an expression of unused `if` branch and then dropped the result, as it was not needed. The exception was raised as a side effect though. My colleague pointed out, that it is allowed to do so, as long as no `FENV_ACCESS` pragma is used. When I tried to enable `FENV_ACCESS` I got following warning from clang:
```
warning: '#pragma FENV_ACCESS' is not supported on this target - ignored [-Wignored-pragmas]
```
We could drop the math_errhandling test as not supported, but it explores many precious corner cases for math functions. Here comes the idea to disable just the exception checks.

The runtime check would not help in this case. Clang would be detected to produce the exceptions, but the test would later fail, because of the spurious exceptions.

As we have at least two factors, which can affect the applicability of exception checks (compiler + runtime library), maybe a build option to control these checks would be a more flexible solution?
```
option('tests-enable-fe-exception-checks', type: 'boolean', value: true,
       description: 'tests enable floating point exceptions checks')
```
Or maybe some combination of build option and compiler checks?

**Comment by keith-packard** [comment]
I've actually fixed some bugs in fenv support when running clang of this nature -- had to rework the code so that clang wouldn't lift computations that might generate exceptions above a conditional branch, like in 659b85d00cbd00d975cc0248d9b28e99ca1d3d26 So, even though clang appears to not reliably provide precise exception support, it is possible to fix the source code so that exceptions are supported correctly by the resulting library. In this case, I think the question is whether we can come up with a run-time detection mechanism that we trust to provide correct answers even on future versions of clang. I think it's possible, using a combination of barriers and volatile variables.

**Comment by keith-packard** [comment]
I've looked at this a bit more closely -- one feature in picolibc's fenv.h is that it sets `PICOLIBC_DOUBLE_NOEXCEPT` and `PICOLIBC_LONG_DOUBLE_NOEXCEPT` values. Those can't be computed at runtime, so we actually do need to know at build time and the only way I know how to do this is by having each architecture set those variables correctly.

So, what I've done in #503 is to validate those settings in the `math_errhandling` test and not changing anything else. I think that's about the best we can do.

**Comment by keith-packard** [comment]
I found something today that I thought was interesting -- the soft FP code for risc-v *does* support exceptions on hardware that has the necessary registers. For instance, long double soft float functions raises exceptions when run on hardware with only 32-bit float support, but does not raise exceptions on hardware with no float support. This seems like the right behavior to me, and I kinda wonder if libgcc should be "fixed" to do this on ARM as well. Looking at the picolibc code, only ARM sets any of the `PICOLIBC_*_NOEXCEPT` values.

**Comment by domin144** [comment]
I read your pull request https://github.com/picolibc/picolibc/pull/503 . It prevents as from setting incorrect `PICOLIBC_*_NOEXCEPT`. It does not solve the problem I face, though.

The case I analyse is https://github.com/domin144/LLVM-embedded-toolchain-for-Arm, platforms `armv7em_hard_fpv4_sp_d16` and `armv7em_hard_fpv5_d16`. The compiler/platform combination I deal with produces the exceptions, but not the way the math_errhandling test expects it to do. These are the possible solutions I can see:
a) Fix the library to produce the expected exceptions.
b) give an option to only disable exception checks, but still check the values <- this is what I propose in this PR.
c) disable math_errhandling test completely - this is what I do for now, but this is are real loss not to check at least the values of the results. I did not see these corner cases checked anywhere else.

The option a) is the nicest of them, but I think it's a long term task and very hard to achieve reliably without compiler supporting the `FENV_ACCESS` directive. I tried to apply the approach you took in https://github.com/picolibc/picolibc/commit/659b85d00cbd00d975cc0248d9b28e99ca1d3d26 to atan2_nottiny test case for `double`.

In this fragment of atan264 function:
```
    /* compute y/x */
    k = (iy - ix) >> 20;
    if (k > 60)
        z = pi_o_2 + _F_64(0.5) * pi_lo; /* |y/x| >  2**60 */
    else if (hx < 0 && k < -60)
        z = _F_64(0.0); /* |y|/x < -2**60 */
    else
        z = atan64(check_uflow(fabs64(y / x))); /* safe to do y/x */
    switch (m) {
    case 0:
        return z; /* atan(+,+) */
    case 1: {
        __uint32_t zh;
        GET_HIGH_WORD(zh, z);
        SET_HIGH_WORD(z, zh ^ 0x80000000);
    }
        return z; /* atan(-,+) */
    case 2:
        return pi - (z - pi_lo); /* atan(+,-) */
    default: /* case 3 */
        return (z - pi_lo) - pi; /* atan(-,-) */
    }
```
for this particular test case `hx < 0`, `m == 2`, `k == -1006`, so `z = _F_64(0.0);`. My current understanding is (I am not fluent in asm, so I cannot confirm this) that GCC takes this as a special case and returns precalculated value `pi - (0.0 - pi_lo)` without raising the "inexact" exception. I managed to make clang not to emit the exception by replacing the `if` branch with following precalculated values:
```
    else if (hx < 0 && k < -60)
    {
        switch (m) {
        case 0: /* atan(+,+) */
        case 1: /* atan(-,+) */
            return _F_64(0.0);
        case 2: /* atan(+,-) */
            return pi;
        default: /* atan(-,-) */
            return -pi;
        }
    }
```

 I have a couple of questions now.
1. Is this explanation plausible?
2. Is this acceptable and intended to rely on GCC to make the optimization?
3. Where do the expectations expressed in the test come from? The C standard seems to leave it unspecified, whether `atan2` should raise "inexact" exception. If one tried to follow literally the "inexact" exception definition "result is not representable exactly" (by wikipedia https://en.wikipedia.org/wiki/IEEE_754#Exception_handling), shouldn't the "inexact" exception be raised in this case? `π` is not exact result in this case, and even if it was,  `π` cannot be represented exactly in any floating point type.

**Comment by keith-packard** [comment]
> for this particular test case `hx < 0`, `m == 2`, `k == -1006`, so `z = _F_64(0.0);`. My current understanding is (I am not fluent in asm, so I cannot confirm this) that GCC takes this as a special case and returns precalculated value `pi - (0.0 - pi_lo)` without raising the "inexact" exception. 

No, gcc actually does both subtractions at runtime for the thumb_v7e_m_dp_hard target (I think that's the gcc equivalent for one of the targets you're using). However, this won't generate an exception because the result doesn't depend on the precision used during the computation. Here's the relevant bit from the IEEE 754 standard (it appears that wikipedia mis-quotes this bit):

>   7.6 Inexact 7.60
>
> Except as specified otherwise (e.g., 5.8 and 5.9), an operation delivering a numerical result that signals no
>  other exception shall signal inexact if its rounded result differs from what would have been computed were
>  both exponent range and precision unbounded. The rounded result shall be delivered to the destination.

Specifically, you don't get Inexact just because the actual value cannot fit in the available precision, but only when the computation was affected by the limited precision of intermediate values. In this case, the result wasn't affected by the precision used during the computation; the only intermediate value is `0.0 - pi_lo` and that is computed exactly.

>     2. Is this acceptable and intended to rely on GCC to make the optimization?

I haven't found any cases where GCC doesn't generate precise exceptions while I've found many cases where clang does not. In this case, GCC performs exactly the operations specified in the code and the hardware does not generate an Inexact exception because the computation shouldn't raise one.

>     3. Where do the expectations expressed in the test come from? The C standard seems to leave it unspecified, whether `atan2` should raise "inexact" exception.

They do come from the IEEE specification -- in this case, because the value of y/x is so close to zero, the computation of atan2(y,x) can be computed exactly (i.e., as if there were no limit on the precision used during the computation). And the atan2 implementation in picolibc reflects that -- when `|y/x| < 0x1p-60`, then the library "knows" the result is exact.

> If one tried to follow literally the "inexact" exception definition "result is not representable exactly" (by wikipedia https://en.wikipedia.org/wiki/IEEE_754#Exception_handling), shouldn't the "inexact" exception be raised in this case? `π` is not exact result in this case, and even if it was,  `π` cannot be represented exactly in any floating point type.

It looks like wikipedia is simplifying what 'inexact' means here -- it's a very subtle distinction, but the inexact exception doesn't say the 'true result' had to be rounded, rather it says that the resulting computation was affected by the limits of precision (or exponent range). In this case, the result was unaffected by limited precision and hence no Inexact exception should be raised.

Thanks for helping me learn about the `#pragma STDC FENV_ACCESS ON` stuff though; it's pretty clear that both gcc and clang should add support for this. In the case of GCC (which appears to act as if the default were "ON"), you could disable this and improve performance. In the case of clang (which appears to act as if the default were "OFF"), you could enable this and make the math library work correctly without struggle.

In any case, now we need to decide what to do for this case when using clang as that compiler takes a very casual approach to floating point exceptions. Have you managed to reproduce this with a target that is supported by qemu? If so, we can add that target to the existing CI setup and fix the bugs present by inserting `clang_barrier_double` where necessary to force evaluation order.

**Comment by keith-packard** [comment]
Ok, I just ran the math errhandling tests under the current debian verison of clang on the thumbv7e+fp platform and found two errors in `nexttowardf` caused by some missing barriers needed to get clang to actually perform the operations necessary to generate the `FE_UNDERFLOW` exception. This is a new regression with version 14.0.6 as I hadn't seen that in the past, and this is tested in CI which is running debian testing. I've pushed #504 with fixes for those. Let's get the rest of the bugs you've found fixed instead of working around them by ignoring the problem.

**Comment by domin144** [comment]
You are completely right about the `FE_INEXACT` masking! I made a silly mistake at the very beginning of the investigation. In my case, I have `PICOLIBC_DOUBLE_NOEXCEPT` defined and the exception is still generated for the `double` version of `atan2`.
I need to take a closer look at what happened here.

**Comment by keith-packard** [comment]
> You are completely right about the `FE_INEXACT` masking! I made a silly mistake at the very beginning of the investigation. In my case, I have `PICOLIBC_DOUBLE_NOEXCEPT` defined and the exception is still generated for the `double` version of `atan2`. I need to take a closer look at what happened here.

If your platform has hardware double support, then `PICOLIBC_DOUBLE_NOEXCEPT` shouldn't be getting set. Check out the code which selects `PICOLIBC_DOUBLE_NOEXCEPT` around https://github.com/picolibc/picolibc/blob/31ff1b3601b379e4cab63837f253f59729ce1fef/newlib/libc/machine/arm/sys/fenv.h#L44 -- that's looking at the value of `__ARM_FP` which is documented here: https://developer.arm.com/documentation/dui0774/g/chr1383660321827

In short -- on 32-bit ARM, soft float and double support doesn't generate exceptions. On targets with hard float and soft double, the double paths shouldn't ever generate exceptions. There could be bugs in picolibc which end up generating spurious exceptions, or failing to generate exceptions, because the code has lots of places which perform computations to intentionally generate the correct exceptions even when the values from those computations isn't relevant.

It sounds like there are just bugs somewhere on your platform, and that we can probably fix them.

**Comment by domin144** [comment]
I found, that compiler-rt produces the *inexact* exception for `armv7em_hard_fpv4_sp_d16` platform in `__aeabi_dadd`, but does not produce *any* exceptions in `__aeabi_ddiv`.
Disabling this line: https://github.com/llvm/llvm-project/blob/dac5957ca934206f4b0777e62f093b08919e36a9/compiler-rt/lib/builtins/fp_add_impl.inc#L170 fixes all double and long double tests for this platform. I will have to consult this with compiler-rt authors.
Many thanks for the patience and support.
I will close this PR now, as it does no longer seem to be needed.

**Comment by keith-packard** [comment]
Thanks for finding the root cause of this bug -- this process did uncover bugs in the picolibc exception support for clang, which is also good.
